### ISSUE TITLE
the model weight seems not been updated

### ISSUE THREAD

**Issue by henryxiao1997** [OpenLMLab/LOMO#14]
When i evalutae it on the dev dataset during training iterations, the performance (PPL) does not change after training one epoch. When I look into the trainer code, there seems no code to update the model weigth. Only the code to update the optimizer. Is it expected (only update optimizer but not weight  --- i.e. loss.backword)? Thanks!

**Comment by KaiLv69** [comment]
Hi. The weight is updated during backward. The implementation is based on hook function in PyTorch. The `Implementation` Section in `README` provides a more detailed introduction to this process. 
You can also check the implementation of the code here.
1. Register hook function: 
https://github.com/OpenLMLab/LOMO/blob/548d8e0af5ce73c467c39c7e1471f3ba6b8b8785/src/lomo.py#LL60-L62
2. Hook function:
https://github.com/OpenLMLab/LOMO/blob/548d8e0af5ce73c467c39c7e1471f3ba6b8b8785/src/lomo.py#LL107-L152

**Comment by henryxiao1997** [comment]
> Hi. The weight is updated during backward. The implementation is based on hook function in PyTorch. The `Implementation` Section in `README` provides a more detailed introduction to this process. You can also check the implementation of the code here.
> 
> 1. Register hook function:
>    https://github.com/OpenLMLab/LOMO/blob/548d8e0af5ce73c467c39c7e1471f3ba6b8b8785/src/lomo.py#LL60-L62
> 2. Hook function:
>    https://github.com/OpenLMLab/LOMO/blob/548d8e0af5ce73c467c39c7e1471f3ba6b8b8785/src/lomo.py#LL107-L152

Got it! Thanks for your quick reply!

So, where and when are these weights updated? Is it in this line (https://github.com/OpenLMLab/LOMO/blob/548d8e0af5ce73c467c39c7e1471f3ba6b8b8785/src/lomo_trainer.py#L173), or in the early this one (https://github.com/OpenLMLab/LOMO/blob/548d8e0af5ce73c467c39c7e1471f3ba6b8b8785/src/lomo_trainer.py#L159, or both of them?

Thanks for your patience!

**Comment by KaiLv69** [comment]
The weights are updated in `self.optimizer.fused_backward()`. The `.get_param_coordinator().reset_step()` is a function in DeepSpeed to indicate that we have completed one fwd+bwd for the model.

**Comment by henryxiao1997** [comment]
Understood. Thank you very much!
